### DISCUSSION TITLE
[General] Account Abstraction: False Positive Transfers on Custom Accounts

### DISCUSSION THREAD

**Discussion by gauravl-tevaeralabs** [zkSync-Community-Hub/zksync-developers#158] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

We're trying to implement a session key by extending the IAccount interface to create a session wallet.

### What is the specific issue or error you're encountering?

For custom accounts established through Account Abstraction, transfers appear to be processed successfully to any specified wallet; however, no actual movement of ETH occurs. Despite transactions appearing successful on the explorer, no deduction of ETH is evident from the custom account, and the recipient does not receive any funds. Below is a sample transaction illustrating this issue. In this specific transaction, the sender utilized a custom account to transfer 0.005ETH, but no actual transfer of ETH transpired. Consequently, the balances of both the sender and the receiver remain unchanged:

https://explorer.zksync.io/tx/0xcb861de61fb7877b213b248772f010f47d4f229c2fb9cfd06a1b3d548ee5413b

### Can you share the error messages or logs you're receiving, if any?

NA

### Have you made any recent changes to the contract before encountering this issue?

No

### Are there any external libraries or contracts that your contract interacts with?

No

### Can you provide the relevant portions of your contract code where the issue is occurring?
```
// SPDX-License-Identifier: BUSL-1.1
pragma solidity ^0.8.18;

import "@matterlabs/zksync-contracts/l2/system-contracts/interfaces/IAccount.sol";
import "@matterlabs/zksync-contracts/l2/system-contracts/libraries/TransactionHelper.sol";
import "@openzeppelin/contracts/access/Ownable.sol";
import "@openzeppelin/contracts/interfaces/IERC1271.sol";
import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
import "@matterlabs/zksync-contracts/l2/system-contracts/Constants.sol";
import "@matterlabs/zksync-contracts/l2/system-contracts/libraries/SystemContractsCaller.sol";

import "./Session.sol";

contract SessionAccount is Ownable, Session, IAccount {
    using TransactionHelper for Transaction;

    bytes4 constant EIP1271_SUCCESS_RETURN_VALUE = 0x1626ba7e;

    modifier onlyBootloader() {
        require(
            msg.sender == BOOTLOADER_FORMAL_ADDRESS,
            "Only bootloader can call this method"
        );
        _;
    }

    constructor(address owner) {
        _transferOwnership(owner);
    }

    function validateTransaction(
        bytes32,
        bytes32 _suggestedSignedHash,
        Transaction calldata _transaction
    ) external payable override onlyBootloader returns (bytes4 magic) {
        magic = _validateTransaction(_suggestedSignedHash, _transaction);
    }

    function _validateTransaction(
        bytes32 _suggestedSignedHash,
        Transaction calldata _transaction
    ) internal returns (bytes4 magic) {
        SystemContractsCaller.systemCallWithPropagatedRevert(
            uint32(gasleft()),
            address(NONCE_HOLDER_SYSTEM_CONTRACT),
            0,
            abi.encodeCall(
                INonceHolder.incrementMinNonceIfEquals,
                (_transaction.nonce)
            )
        );

        bytes32 txHash;
        if (_suggestedSignedHash == bytes32(0)) {
            txHash = _transaction.encodeHash();
        } else {
            txHash = _suggestedSignedHash;
        }

        require(
            _transaction.signature.length == 65,
            "Signature length is incorrect"
        );

        address signer = ECDSA.recover(txHash, _transaction.signature);
        address to = address(uint160(_transaction.to));

        require(
            owner() == signer || inSession(signer, to),
            "Signature is invalid"
        );

        magic = ACCOUNT_VALIDATION_SUCCESS_MAGIC;
    }

    function executeTransaction(
        bytes32,
        bytes32,
        Transaction calldata _transaction
    ) external payable override onlyBootloader {
        _executeTransaction(_transaction);
    }

    function _executeTransaction(Transaction calldata _transaction) internal {
        address to = address(uint160(_transaction.to));
        uint256 value = _transaction.reserved[1];
        bytes memory data = _transaction.data;

        if (to == address(DEPLOYER_SYSTEM_CONTRACT)) {
            SystemContractsCaller.systemCall(
                uint32(gasleft()),
                to,
                uint128(_transaction.reserved[1]),
                _transaction.data
            );
        } else {
            bool success;
            assembly {
                success := call(
                    gas(),
                    to,
                    value,
                    add(data, 0x20),
                    mload(data),
                    0,
                    0
                )
            }
            require(success);
        }
    }

    function executeTransactionFromOutside(
        Transaction calldata _transaction
    ) external payable {
        _validateTransaction(bytes32(0), _transaction);

        _executeTransaction(_transaction);
    }

    function payForTransaction(
        bytes32,
        bytes32,
        Transaction calldata _transaction
    ) external payable override onlyBootloader {
        bool success = _transaction.payToTheBootloader();
        require(success, "Failed to pay the fee to the operator");
    }

    function prepareForPaymaster(
        bytes32,
        bytes32,
        Transaction calldata _transaction
    ) external payable override onlyBootloader {
        _transaction.processPaymasterInput();
    }

    receive() external payable {
        assert(msg.sender != BOOTLOADER_FORMAL_ADDRESS);
    }
}
```
### Have you tried to isolate the problem, and if so, what were the results?

I have attempted to isolate the problem by creating a session wallet and by implementing the spending limits example provided by the team on GitHub. However, the same issue persists. Given that this scenario presents no error messages or outputs, troubleshooting has proven to be challenging.

### What steps have you already taken to try to resolve the issue?

To address the issue, I attempted to replace SystemContractsCaller.systemCall() with SystemContractsCaller.systemCallWithPropagatedRevert(), and also tested it using the latest version of zksol. Additionally, I referred to the zksync examples provided for account abstraction and cross-verified with those examples to ensure I haven't overlooked any aspects.

### Repo Link (Optional)

https://github.com/tevaera-labs/contracts/blob/main/contracts/teva-system/SessionAccount.sol

**Comment by ljankovic-txfusion** [comment, marked as answer]
Check https://github.com/zkSync-Community-Hub/zkync-developers/discussions/165#discussioncomment-7210554 , it seems to be the cause of both of these issues. The `msg.value` is not passed correctly.

**Reply by gauravl-tevaeralabs** [reply]
Thank you, @ljankovic-txfusion! It resolves the issue.

**Comment by augustinov** [comment]
It appears that the issue might be related to the execution of transactions within the executeTransaction function. Please ensure that the transaction execution logic within this function is correctly implemented, especially when interacting with external contracts.

Additionally, double-check the validation and execution process within the validateTransaction and executeTransaction functions respectively, to ensure that the transaction parameters are handled properly and the validation criteria are met.

Consider adding additional logging or debugging statements within these functions to gain more insights into the transaction flow and identify any potential issues.

If the problem persists, it might be beneficial to review the transaction lifecycle and interaction with external contracts to identify any discrepancies or errors in the logic.

Lastly, consider reaching out to the zkSync community or support channels for further assistance and troubleshooting.

**Comment by backvomeph** [comment]
Ensure that the transaction validation process in the SessionAccount contract is correctly configured to verify signatures and execute transactions. Review the interaction with external contracts and confirm that the necessary calls are made to facilitate the transfer of ETH. Check for any discrepancies in nonce handling or signature validation that may be causing the issue.

**Comment by dunminecom** [comment]
Given the complexity of the interaction and the absence of error messages, further debugging may require in-depth analysis of the contract logic, including potential issues with transaction validation, execution, and interaction with other system contracts. Reviewing the entire transaction lifecycle and comparing it with working examples from the ZKSync repository could help identify the root cause of the issue.
